### PR TITLE
feat(paper-index): NFL deserved-win fit

### PR DESCRIPTION
**Pull request by saiemgilani**
Ships the NFL Paper Index (Deserved Win %): weights fitted by `python/tools/fit_paper_index.py --league nfl` on the espn_nfl play-by-play rebuilt with `scoring_opp` keyed to `start.yardsToEndzone` and nflverse closing lines (sportsdataverse-py #495). `paper_index.compute(..., league="nfl")` now returns a share, so the NFL game page renders the panel.

Evidence routes: /game/401856682 /nfl/game/401772945

## Fit: two splits, decided once and disclosed

The trainer now gates on a **paired** rule against the EPA-only baseline: per holdout game d = (p − y)² − (p_epa − y)², require mean(d) + 2·se ≤ 0 (1,000-draw bootstrap CI recorded). The original 2024-25 holdout beat EPA-only in the point estimate but **failed** that rule — underpowered at n=570 — so the holdout was widened for power, one time, to 2022-2025. Both runs (Pro Bowls dropped, FG points counted):

| run | train / holdout | n hold | Brier | EPA-only | paired mean(d) | se | mean+2se | bootstrap 95% | log-loss | acc | reliab. | resol. | pinned |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| A (first choice) | 2016-23 / 2024-25 | 570 | 0.1182 | 0.1322 | −0.0141 | 0.0078 | **+0.0016 FAIL** | [−0.0299, +0.0010] | 0.3698 | 0.8263 | 0.0029 | 0.1317 | explosive |
| **B (shipped)** | 2016-21 / 2022-25 | 1,137 | **0.1174** | 0.1374 | −0.0200 | 0.0054 | **−0.0092 pass** | [−0.0310, −0.0100] | 0.3677 | 0.8276 | 0.0026 | 0.1316 | none |
| college, for scale | 2016-23 / 2024-25 | 1,894 | 0.0657 | 0.0805 | −0.0147 | 0.0039 | −0.0069 pass | [−0.0223, −0.0069] | 0.2172 | 0.9108 | 0.0011 | 0.1660 | none |

Earlier diagnostics on split A: no FG points 0.1198; pre-#495 data (wrong-yardline `scoring_opp`, league pts/opp 0.99) lost outright, 0.1373 vs 0.1322 with both opportunity margins pinned. Coin-flip Brier 0.25.

Shipped NFL weights (`WEIGHTS["nfl"]`): success 8.0168, explosive 1.3517, explosive_epa 1.7687, opp_conversion 1.9426, pts_per_opp 0.3468, field_position 2.5097, havoc 5.7830, turnovers 0.4545; `LEAGUE_PTS_PER_OPP["nfl"] = 3.7306` (the no-opportunity neutral, from the train seasons' fitted games). The same rule moves college's constant 3.3566 → 3.3603 with every CFB weight unchanged to four decimals. Gates pinned from the observed values: `GATES["nfl"] = {min_holdout: 1000, brier: 0.13, resolution: 0.10, reliability: 0.01}` + the paired rule.

Calibration (holdout B, 10 bins, n / mean p / mean y): 216 / .039 / .069 · 101 / .151 / .139 · 70 / .251 / .271 · 76 / .352 / .395 · 74 / .452 / .459 · 72 / .556 / .486 · 74 / .645 / .689 · 88 / .755 / .898 · 109 / .853 / .872 · 257 / .964 / .996.

### Field-goal points in points-per-opportunity (NFL only)
A made field goal is a non-scrimmage row in both leagues' pbp, so the scrimmage filter dropped its 3 points. The NFL fit counts made field goals flagged `scoring_opp` (`paper_index.OPP_POINTS_INCLUDE_FG = {"cfb": False, "nfl": True}`); train and serve build the feature through one shared `opp_points_mask(league)`, and the trainer's parity check compares `team_inputs()` against its own aggregation on real holdout games. College stays as fitted; the CFB fit reproduces its committed weights and shares exactly through the changed trainer.

### Holdout contamination, measured
The bundled NFL field-position curve is fitted on 2016-2025 drives and the NFL EP model on 1999-2025 (college: curve 2018-2021, EP model 2004-2025 per its card) — recorded in `provenance.holdout_contamination`. `tools/paper_index_curve_sensitivity.py` refits the curve on the train seasons only (max shift 0.28 EP) and re-runs the fit: holdout Brier 0.1177 vs 0.1174 with the bundled curve.

### sdv-model-reviewer audit items (all landed)
1. Curve fingerprint (`sha256` + EP at yardlines 1/20/50/80/99) written to provenance and asserted per league against the installed sportsdataverse.
2. Gates re-asserted in pytest from the committed provenance (floors, point and paired EPA-only rules, bootstrap upper bound ≤ 0).
3. Provenance records pbp source, per-season input identity (local size+sha256 / release asset size+updated_at), per-season row accounting, sportsdataverse version + git sha, fitted_at.
4. Cache keyed by input identity (+ curve sha256 + FG flag for ext rows), not by path; `--refresh` forces.
5. Paired delta with se and bootstrap CI; gate `mean+2se ≤ 0`.
6. Contamination documented in docstring + provenance; sensitivity script committed.
7. `game_id` dtype asserted; every inner join must keep ≥98% of its left side; per-season counts recorded.
8. Franchise allowlist {1..30, 33, 34}: Pro Bowls 400878708, 400999088, 401038851, 401131046, 401326637 dropped and listed in provenance.
9. `python/tests/fixtures/README.md`.
10. Model registry rows + retrain runbook in `CLAUDE.md`.
11. 10-bin calibration table on stdout and in provenance.
12. KKT re-entry unit-tested on a synthetic collinear design (one-way active set pins {0,1}; KKT ends at {1,3}, higher likelihood, non-positive gradient on pinned columns).
13. No-opportunity `ptsPerOpp` fill asserted for both leagues.

Copilot's later passes (3cd7005a, b54ff3a7, and the ppo fix): league pts/opp derived from the fitted games only (Pro Bowls and filtered games excluded — this moved the NFL constant 3.7257→3.7306 and college's 3.3566→3.3603, weights unchanged beyond the fourth decimal); cache key by input identity (not path) with a hash of every row-building helper; empty active set handled; local-path guard; sensitivity probe in scratch caches; holdout panel sorted before scoring so the seeded bootstrap is deterministic (the CI in the fixture and here now agree); gates read full-precision metrics and also enforce the bootstrap upper bound; cache keys fold in a hash of the aggregation source/columns/filters so a feature change invalidates rows without `--refresh`; release-asset map fetched once per league; keep-floor assert on the drives join; test docstring states each league's holdout seasons.

### Other changes
- `share_from_inputs` reports only fitted (weight > 0) margins.
- `python/uv.lock`: sportsdataverse pinned to sdv-py main `8dbbfa2e` (was `af3150fe`, pre-#495): the live processor must compute `scoring_opp` the way the training data does, and it ships the bundled `nfl_field_position_ep.parquet`.
- `PaperIndex.astro`: team logos follow `espnLogoLeague(Astro.locals.league)`; asserted in `nflGamePage.render.test.ts`.

## Evidence: the flagged HTML-transfer growth is the sdv-py payload, not the panel
The evidence workflow runs each tree's backend from its own lock (base sdv-py `af3150fe`, head `8dbbfa2e`). Measured by calling both backends' `/process` routes and gzipping each top-level key:

| route | base payload | head payload | where it grew |
|---|---|---|---|
| `/game/401856682` (CFB) | 273 KB gz | 304 KB gz | `advBoxScoreSpans` +23.6 KB, `advBoxScore` +5.7 KB, situational +2.7 KB (sdv-py #496); DOM 14,075 → 14,078 |
| `/nfl/game/401772945` | 180 KB gz | 300 KB gz | base NFL processor exposed **no `plays_frame`**; head adds `advBoxScoreSpans` +60 KB, `plays` +37 KB, `advBoxScore` +8 KB, situational/drive spans +14 KB, `paperIndex` **+0.7 KB**; DOM +299 = the new blocks |

## Verification
- `cd python && PYTHON_HTTP_TOKEN=test uv run pytest -q`: 65 passed.
- `astro` vitest (docker node:22): 37 files, 319 passed, 1 skipped.
- Trainer: NFL and CFB runs exit 0 with parity check OK; CFB weights and fixture games byte-identical to the committed fit.

## Summary by Sourcery

Ship a league-specific NFL Paper Index model with validated training provenance and display its deserved-win shares on NFL game pages.

New Features:
- Add NFL Paper Index (Deserved Win %) scoring with league-specific fitted weights and neutral points-per-opportunity values.
- Render the Paper Index panel on NFL game pages with NFL team logos and quarter-level shares.

Bug Fixes:
- Count made NFL field goals in points-per-opportunity calculations while preserving the existing college behavior.
- Prevent upstream field-position curve changes from silently altering served shares by recording and validating curve fingerprints.

Enhancements:
- Expand Paper Index computation and training to support independent league models, league-specific expected-points curves, and fitted-factor reporting.
- Strengthen model fitting with non-negative active-set optimization, paired EPA-only evaluation, bootstrap confidence intervals, calibration metrics, provenance, contamination tracking, and input-integrity checks.
- Add content- and feature-aware training caches, retraining guidance, model registry documentation, and a curve-sensitivity analysis tool.

Build:
- Pin sportsdataverse to the revision containing the NFL play-by-play processing and field-position model updates.

Documentation:
- Document Paper Index model artifacts, training data, gates, provenance, and the retraining runbook.
- Add documentation for Paper Index oracle fixtures, data sources, row accounting, known gaps, and holdout contamination.

Tests:
- Add NFL oracle fixtures and cross-league tests covering model parity, gates, calibration provenance, field-goal handling, curve identity, non-negative fitting, and rendered NFL logos.

### CODE REVIEW

File: python/tools/fit_paper_index.py [hunk 1]
```diff
@@ -27,7 +27,27 @@
 shipped spec: Brier 0.0657, 91.1% winner agreement, all weights positive.
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply python/tools/fit_paper_index.py]: Fixed in 1700775: the docstring now says the fit is non-negative and that a margin turning negative under collinearity is pinned to zero (the NFL's explosive-play rate), rather than claiming every shipped weight is positive.

**Comment by saiemgilani** [comment]
Follow-up to the Copilot review's four suppressed findings (1700775):

**Non-negative fit is not a complete active-set solve (fit_paper_index.py:391)** — agreed in general. `fit_logistic_nonneg` now re-enters a pinned margin when the log-likelihood gradient along it at the current solution is positive (the KKT condition for a bound at zero), and stops only when every active weight is positive and every pinned margin has a non-positive gradient; the objective is concave so that is the constrained optimum. Both league fits reproduce exactly: the NFL's pinned explosive-play rate has a non-positive gradient, college pins nothing.

**Cache keyed only by league + mtime (fit_paper_index.py:128)** — the per-season cache now lives under `.paper_index_cache/<league>/<sha1 of the pbp source>[:10]`, so rows aggregated from one `--pbp-dir` never serve a fit pointed at another; the mtime check still rebuilds when a local parquet is newer, and `--refresh` forces it.

**Holdout-era drives in the field-position EP curve (fit_paper_index.py:292)** — measured rather than assumed. The bundled NFL curve is fitted on 2016-2025 drives. Refitting it with `fit_nfl_field_position_ep` on 2016-2023 drives only moves it by at most 0.14 EP (mean 0.04), and re-running the trainer against that curve gives holdout Brier 0.1181 vs 0.1183 with the bundled curve (weights within 0.07 of the shipped ones; the same margin pins). The shared curve does not flatter the evaluation; documented in the trainer docstring. Separating a validation-only curve from the production artifact in sdv-py is not worth the second artifact for a 0.0002 difference.

**Render test injects `game.paperIndex` (nflGamePage.render.test.ts:104)** — the committed processed-game fixture predates the NFL fit, and the injection is deliberate so the test exercises the template rather than a 2 MB fixture refresh. The backend path is covered on the Python side: `test_compute_is_none_for_a_league_without_a_fit` asserts `compute(..., league="nfl")` returns a share, `test_field_goal_points_count_per_league` covers the fail-open when `fg_made` is missing (the live NFL processor emits it: `type.text == "Field Goal Good"`), and the trainer's parity check runs `team_inputs()` on real 2024 pbp against its own aggregation. The pr-evidence workflow renders `/nfl/game/401772945` end to end for this PR.

**Comment by saiemgilani** [comment]
Evidence note: the flagged HTML-transfer growth on both routes is the sdv-py lock bump's larger processed-game payload (CFB: +32 KB gz in advBoxScoreSpans/advBoxScore from #496 with an unchanged DOM; NFL: the base backend exposed no plays_frame, so span boxes, drive summary, situational stats and the Paper Index are all new). The Paper Index object itself is 706 B gz. Measured per top-level key against both backends; table added to the PR description.

**Comment by saiemgilani** [comment]
Audit follow-up landed at 60bae226 (all 13 sdv-model-reviewer items; PR description rewritten with both NFL runs). The 2024-25 holdout failed the new paired EPA-only gate (mean -0.0141, se 0.0078, mean+2se +0.0016); the shipped fit is the 2016-21 / 2022-25 split (n=1,137): Brier 0.1174 vs 0.1374, paired -0.0200 ± 0.0054, mean+2se -0.0092, every weight positive.

File: python/tests/fixtures/paper_index_oracle_nfl.json [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+  "holdout_reliability": 0.0026,
+  "holdout_uncertainty": 0.2473,
+  "epa_only_brier": 0.1374,
+  "epa_only_paired": {
+   "n": 1137,
+   "mean_delta": -0.02,
+   "se": 0.0054,
+   "ci95": [
+    -0.0329,
+    -0.0088
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply python/tests/fixtures/paper_index_oracle_nfl.json]: Reconciled in 3cd7005a. The mismatch was real and had a cause: the holdout panel came out of polars hash joins in thread-dependent order, so the seeded bootstrap drew differently per run. build_games() now sorts the panel by (season, game_id) before scoring; two consecutive runs agree (NFL CI [-0.0310, -0.0100], CFB [-0.0223, -0.0069]) and the PR description carries those values.

File: python/tools/fit_paper_index.py [hunk 3]
```diff
@@ -24,42 +24,254 @@
...
+    src = PBP_URL.format(season=season)
+    path = pathlib.Path(src)
+    if path.exists():
+        h = hashlib.sha256()
+        with path.open("rb") as fh:
+            for chunk in iter(lambda: fh.read(1 << 20), b""):
+                h.update(chunk)
+        ident = {"source": src, "bytes": path.stat().st_size, "sha256": h.hexdigest()}
+    else:
+        asset = _release_assets(LEAGUE).get(f"play_by_play_{season}.parquet")
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply python/tools/fit_paper_index.py:263]: Fixed in 3cd7005a: _release_assets() memoizes the asset map per league in a module-level dict, so a ten-season fit makes one API call.

File: python/tools/fit_paper_index.py [hunk 4]
```diff
@@ -202,57 +443,90 @@ def season_ext_rows(season: int) -> pl.DataFrame:
...
+    # fitted constant, and a cache must not bake in whichever value the
+    # module happened to hold when the rows were aggregated
     out = base.join(drives, on=["game_id", "pos_team_id"], how="inner").with_columns(
         opp_conv_rate=pl.when(pl.col("opp_trips") > 0)
         .then(pl.col("opp_converted") / pl.col("opp_trips"))
         .otherwise(0.5),
-        pts_per_opp=pl.when(pl.col("opp_trips") > 0)
-        .then(pl.col("opp_points") / pl.col("opp_trips"))
-        .otherwise(LEAGUE_PTS_PER_OPP),
     )
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply python/tools/fit_paper_index.py:495]: Fixed in 3cd7005a: _assert_join_kept(out, base, ...) now follows the drives join in season_ext_rows, before the rows are cached.

File: python/tools/fit_paper_index.py [hunk 5]
```diff
@@ -287,17 +597,89 @@ def brier_decomposition(p: np.ndarray, y: np.ndarray, bins: int = 10):
...
+            out.append(f"miscalibrated: reliability {prov['holdout_reliability']}")
+    if prov["holdout_brier"] > prov["epa_only_brier"] + 1e-9:
+        out.append(f"lost to EPA-only: {prov['holdout_brier']} vs {prov['epa_only_brier']}")
+    paired = prov["epa_only_paired"]
+    if paired["mean_delta"] + 2 * paired["se"] > 0:
+        out.append(
+            f"EPA-only not beaten at 2 se: mean delta {paired['mean_delta']}, se {paired['se']}, "
+            f"mean + 2se = {paired['mean_delta'] + 2 * paired['se']:.4f} > 0"
+        )
+    return out
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply python/tools/fit_paper_index.py:733]: Fixed in 3cd7005a: gate_failures() also fails when the bootstrap 95% upper bound is above zero, matching the committed-fixture test.

File: python/tools/fit_paper_index.py [hunk 6]
```diff
@@ -308,129 +690,194 @@ def parity_check(games: pl.DataFrame, season: int, n: int = 5) -> None:
...
+        "league_pts_per_opp": league_ppo,
+        "margins_pinned_to_zero": dropped,
+        "holdout_brier": round(float(np.mean((p_hold - yh) ** 2)), 4),
+        "holdout_log_loss": round(float(-np.mean(yh * np.log(pc) + (1 - yh) * np.log(1 - pc))), 4),
+        "holdout_accuracy": round(float(((p_hold > 0.5) == (yh == 1)).mean()), 4),
+        "holdout_resolution": round(res, 4),
+        "holdout_reliability": round(rel, 4),
+        "holdout_uncertainty": round(unc, 4),
+        "epa_only_brier": round(float(np.mean((p_epa - yh) ** 2)), 4),
+        "epa_only_paired": paired_epa_delta(p_hold, p_epa, yh),
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply python/tools/fit_paper_index.py]: Fixed in 3cd7005a: fit_and_evaluate() keeps a full-precision metrics dict that gate_failures() reads; the provenance gets a rounded copy via _rounded().

File: python/tests/test_paper_index.py [hunk 7]
```diff
@@ -1,23 +1,37 @@
 """Paper Index: unit + real-game oracle tests.
 
-The oracle fixture holds 12 real 2024-2025 holdout games (inputs aggregated
-from the released play-by-play, expected shares computed by the trainer at
-full weight precision). The shipped WEIGHTS are rounded to 4 decimals, so the
-share tolerance is 5e-4 -- tight enough that a wrong sign, a dropped margin,
-or an edited weight fails loudly, loose enough for the documented rounding.
+Each league's oracle fixture holds 12 real 2024-2025 holdout games (inputs
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply python/tests/test_paper_index.py]: Fixed in 3cd7005a: the test module docstring now states each league's holdout seasons (college 2024-2025, NFL 2022-2025).

**Comment by saiemgilani** [comment]
Third-pass suppressed findings addressed in b54ff3a7: cache key hashes the input's identity fields only (bytes+sha256 / bytes+updated_at), not its path; the feature-spec hash also covers _cast_ids and _assert_join_kept; fit_logistic_nonneg treats an empty active set as the zero logit (unit-tested: an all-harmful design ends at the all-zero constrained solution); the sensitivity script docstring states it runs build + fit + holdout scoring only. Fits unchanged (NFL Brier 0.1174, CFB 0.0657, fixtures identical apart from provenance).

File: python/tools/fit_paper_index.py [hunk 8]
```diff
@@ -202,57 +476,96 @@ def season_ext_rows(season: int) -> pl.DataFrame:
...
+        per_season[str(s)] = {"input": season_input(s), **counts}
+    games = pl.concat(frames, how="vertical_relaxed")
+    ext = pl.concat(
+        [season_ext_rows(s).with_columns(season=pl.lit(s)) for s in seasons],
+        how="vertical_relaxed",
+    )
+    train_ext = ext.filter(pl.col("season") < HOLDOUT_SEASONS[0])
+    league_ppo = round(
+        float(train_ext.select(pl.col("opp_points").sum() / pl.col("opp_trips").sum()).item()),
+        4,
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply python/tools/fit_paper_index.py]: Fixed in c8f735c1: build_games() now semi-joins the train-season ext rows on the filtered game table before deriving league_ppo, so Pro Bowl offenses (and tied / sub-floor games) no longer shape the neutral. NFL constant 3.7257 -> 3.7306 (two weights move at the fourth decimal, holdout metrics unchanged); the same rule moves college's 3.3566 -> 3.3603 with every weight unchanged to four decimals. Both fixtures regenerated.

**Comment by saiemgilani** [comment]
Fourth-pass suppressed findings, at c8f735c1: a season file missing under --pbp-dir now fails immediately (no fallback to the release asset's identity); the sensitivity probe runs both evaluations in scratch caches. On the evidence note: the pr-evidence workflow's comment (four screenshots + Lighthouse for /game/401856682 and /nfl/game/401772945) is on this PR and re-posted on every push; the PR description's evidence section explains the flagged HTML growth from measurements against both backends.
